doc/neps/ufunc-overrides.rst

```diff
@@ -106,163 +116,193 @@ Returning ``NotImplemented`` to user should not happen. Moreover::
 standard operator override mechanism, and the two do not interact
 directly.
 
-They however have indirect interactions, because NumPy's ``ndarray``
-type implements its binary operations via Ufuncs. Effectively, we have::
-
-    class ndarray(object):
+They have indirect interactions, however, because NumPy's
+:class:`ndarray` type implements its binary operations via Ufuncs.  For
+most numerical classes, the easiest way to override binary operations is
+thus to define ``__array_ufunc__`` and override the corresponding
+Ufunc. The class can then, like :class:`ndarray` itself, define the
+binary operators in terms of Ufuncs. Here, one has to take some care.
+E.g., the simplest implementation would be::
+
+    class ArrayLike(object):
+        def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
+            ...
+            return result
         ...
         def __mul__(self, other):
-            return np.multiply(self, other)
+            return self.__array_ufunc__(np.multiply, '__call__', self, other)
 
-Suppose now we have a second class::
+Suppose now, however, that ``other`` is class that does not know how to
+deal with arrays and ufuncs, but does know how to do multiplication::
 
     class MyObject(object):
-        def __numpy_ufunc__(self, *a, **kw):
-            return "ufunc"
+        __array_ufunc__ = None
         def __mul__(self, other):
             return 1234
         def __rmul__(self, other):
             return 4321
 
 In this case, standard Python override rules combined with the above
-discussion imply::
+discussion would imply::
 
-    a = MyObject()
-    b = np.array([0])
+    mine = MyObject()
+    arr = ArrayLike([0])
 
-    a * b    # == 1234       OK
-    b * a    # == "ufunc"    surprising
+    mine * arr    # == 1234       OK
+    arr * mine    # TypeError     surprising
 
-This is not what would be naively expected, and is therefore somewhat
-undesirable behavior.
+The reason why this would occur is: because ``MyObject`` is not an
+``ArrayLike`` subclass, Python resolves the expression ``arr * mine`` by
+calling first ``arr.__mul__``. In the above implementation, this would
+just call the Ufunc, which would see that ``mine.__array_ufunc__`` is
+:obj:`None` and raise a :exc:`TypeError`. (Note that if ``MyObject``
+is a subclass of :class:`ndarray`, Python calls ``mine.__rmul__`` first.)
 
-The reason why this occurs is: because ``MyObject`` is not an ndarray
-subclass, Python resolves the expression ``b * a`` by calling first
-``b.__mul__``. Since NumPy implements this via an Ufunc, the call is
-forwarded to ``__numpy_ufunc__`` and not to ``__rmul__``.  Note that if
-``MyObject`` is a subclass of ``ndarray``, Python calls ``a.__rmul__``
-first. The issue is therefore that ``__numpy_ufunc__`` implements
-"virtual subclassing" of ndarray behavior, without actual subclassing.
+So, a better implementation of the binary operators would check whether
```

Comment:
I think it would be clearer if the better, i.e., existing solution is put first followed by the explanation as to why it is done that way. Probably worth mentioning that that makes things easier for folks who are familiar with the Python way and don't want to burrow into the innards of `__array_ufunc__`.